packages/http/test/file.test.ts

```diff
@@ -0,0 +1,547 @@
+
+/*
+
+CASES TO TEST
+
+// `"single"` body: uploads file as a JSON structure { filename: <string>,
+// contentType: <string>, contents: <base64> }
+op uploadFileAsJson(@header contentType: "application/json", @body file: File)
+
+// `"file"` body: Uploads file as a binary data stream. Content-Type has contentType and
+// filename cannot be specified (Content-Disposition is a response-only header)
+op uploadFile(@body file: File);
+op uploadFile(...File);
+
+// `"file"` body: Downloads file as binary data stream. Content-Type has contentType and
+// filename comes from Content-Disposition
+op downloadFile(): { @body file: File };
+op downloadFile(): File;
+
+
+// Declares a custom textual file that can either be "application/json" or "application/yaml"
+model SpecFile extends File<Contents = bytes> {
+  contentType: "application/json" | "application/yaml";
+  // We can put HTTP metadata inside the file model as long as we use it with `@bodyRoot` and not
+  // `@body`      
+  @header("x-filename") filename: string;
+}
+
+// ??? with @header on contentType
+op uploadWeird(@bodyRoot file: SpecFile): void;
+
+// `"file"` body: Uploads spec as a file. Content-Type has contentType, x-filename has filename.
+// isText: true
+op uploadSpec(@bodyRoot spec: SpecFile): void;
+
+// `"single"` body: Uploads spec as a JSON object { contentType: <json or yaml>,
+// filename: <string>, contents: <string> }
+op uploadSpec(@header contentType: "application/json", @body spec: SpecFile): void;
+
+op uploadFileMultipart(@multipartBody fields: { file: HttpPart<SpecFile> }): void;
+
+op uploadFileNested(@body data: { file: File }): void;
+
+op uploadFileInPieces(
+  @header contentType: "application/json" | "application/yaml",
+  // filename?
+  @body contents: bytes,
+}: void;
+*/
+
+describe("http: file", () => {
+  it("does not allow instances that improperly provide contents argument", async () => {
+    const { diagnostics } = await compileOperationsFull(`
+      model BadFile1 extends Http.File<string | bytes> {}
+
+      model BadFile2 extends Http.File<"asdf"> {}
+    `);
+
+    expectDiagnostics(diagnostics, [
+      {
+        code: "@typespec/http/http-file-contents-not-scalar",
+        severity: "error",
+        message:
+          "The 'contents' property of the file model must be a scalar type that extends 'string' or 'bytes'. Found 'string | bytes'.",
+      },
+      {
+        code: "@typespec/http/http-file-contents-not-scalar",
+        severity: "error",
+        message:
+          "The 'contents' property of the file model must be a scalar type that extends 'string' or 'bytes'. Found '<value>'.",
+      },
+    ]);
+  });
+
+  it("exact payload upload and download", async () => {
+    const [
+      {
+        parameters: { body: requestBody },
+        responses: [
+          {
+            responses: [{ body: responseBody }],
+          },
+        ],
+      },
+    ] = await getOperations(`
+      op example(...Http.File): Http.File;
```

Comment:
i was thinking where to reflect the interception for filename and content-type headers, will this logic be in http library or be handled in each emitter?

Reply:
The goal is to remove the emitter's interpretation of what is/isn't a file and how the file is represented entirely. The HTTP library tells you "this is a file" and the content-types that the file allows. So if you see a `"single"` body, you _always_ treat the payload as a structured model. If you see a `"file"` body, you _always_ treat the body as a file.

Reply:
Imagine we could include the filename as response header for case `op example(...Http.File): Http.File;`, my question is if we would add a header named  `Content-Disposition` in http response's header in runtime?

Reply:
We don't have a way to do that today, because the filename does not come from the primary value of the `Content-Disposition` header, it comes from the `filename` parameter of that header's value, and this is something the HTTP library can't communicate today.

`Content-Disposition: attachment; filename=<filename>` -- we just don't have a way to say that a property comes from an extra parameter of the `Content-Disposition` header, but we could potentially do so in the future. In any case, because the file body is a special kind of body, emitters will still be able to account for this when the body is a file.

Reply:
Resolving as there's no way to represent the default location of `filename` in typespec/http today and this behavior of special interpretation of filename's location is already required today in multipart and response files. There may be additional design work for us to consider in the future to represent parameters that are located in header ext-params, but there's nothing to do in this PR to support that.